### ISSUE TITLE
v1.18.3 to v1.19.1 upgrade failed - crashing with exec: \"k3s-server\": executable file not found in $PATH

### ISSUE THREAD

**Issue by dnoland1** [rancher/k3s#2298]
**Environmental Info:**
K3s Version: 

1.18.3 & 1.19.1

Node(s) CPU architecture, OS, and Version: 

Linux ip-10-0-7-10 5.3.0-1035-aws #37-Ubuntu SMP Sun Sep 6 01:17:09 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux

Cluster Configuration:

1 master

**Describe the bug:**

Attempted an upgrade from k3s v1.18.3 to v1.19.1. After the upgrade was run, k3s was continuously crashing with fatal "executable file not found..." error.

**Steps To Reproduce:**

1. Installed k3s 1.18.3 on a fresh Ubuntu 18.04 VM on AWS EC2:
```
# curl -sfL https://get.k3s.io | INSTALL_K3S_VERSION=v1.18.3+k3s1 sh -
[INFO]  Using v1.18.3 as release
[INFO]  Downloading hash https://github.com/rancher/k3s/releases/download/v1.18.3/sha256sum-amd64.txt
root@ip-10-0-7-10:~# curl https://github.com/rancher/k3s/releases/download/v1.18.3/sha256sum-amd64.txt
Not Foundroot@ip-10-0-7-10:~#
root@ip-10-0-7-10:~#
root@ip-10-0-7-10:~# curl -sfL https://get.k3s.io | INSTALL_K3S_VERSION=v1.18.3+k3s1 sh -
[INFO]  Using v1.18.3+k3s1 as release
[INFO]  Downloading hash https://github.com/rancher/k3s/releases/download/v1.18.3+k3s1/sha256sum-amd64.txt
[INFO]  Downloading binary https://github.com/rancher/k3s/releases/download/v1.18.3+k3s1/k3s
[INFO]  Verifying binary download
[INFO]  Installing k3s to /usr/local/bin/k3s
[INFO]  Creating /usr/local/bin/kubectl symlink to k3s
[INFO]  Creating /usr/local/bin/crictl symlink to k3s
[INFO]  Creating /usr/local/bin/ctr symlink to k3s
[INFO]  Creating killall script /usr/local/bin/k3s-killall.sh
[INFO]  Creating uninstall script /usr/local/bin/k3s-uninstall.sh
[INFO]  env: Creating environment file /etc/systemd/system/k3s.service.env
[INFO]  systemd: Creating service file /etc/systemd/system/k3s.service
[INFO]  systemd: Enabling k3s unit
Created symlink /etc/systemd/system/multi-user.target.wants/k3s.service → /etc/systemd/system/k3s.service.
[INFO]  systemd: Starting k3s
```
2. Did a few sanity checks to make sure cluster was healthy:
```
# kubectl get nodes
NAME           STATUS   ROLES    AGE     VERSION
ip-10-0-7-10   Ready    master   6m27s   v1.18.3+k3s1
# kubectl get pods -A
NAMESPACE     NAME                                     READY   STATUS      RESTARTS   AGE
kube-system   local-path-provisioner-6d59f47c7-cdrfm   1/1     Running     0          8m27s
kube-system   metrics-server-7566d596c8-7x2v4          1/1     Running     0          8m27s
kube-system   helm-install-traefik-j9nxw               0/1     Completed   1          8m27s
kube-system   svclb-traefik-pmvcq                      2/2     Running     0          8m6s
kube-system   coredns-8655855d6-bf8tj                  1/1     Running     0          8m27s
kube-system   traefik-758cd5fc85-lfp9b                 1/1     Running     0          8m6s
```
3. After the cluster had been running a little over 10 minutes, performed an upgrade:
```
# curl -sfL https://get.k3s.io | INSTALL_K3S_VERSION=v1.19.1+k3s1 sh -
[INFO]  Using v1.19.1+k3s1 as release
[INFO]  Downloading hash https://github.com/rancher/k3s/releases/download/v1.19.1+k3s1/sha256sum-amd64.txt
[INFO]  Downloading binary https://github.com/rancher/k3s/releases/download/v1.19.1+k3s1/k3s
[INFO]  Verifying binary download
[INFO]  Installing k3s to /usr/local/bin/k3s
[INFO]  Skipping /usr/local/bin/kubectl symlink to k3s, already exists
[INFO]  Skipping /usr/local/bin/crictl symlink to k3s, already exists
[INFO]  Skipping /usr/local/bin/ctr symlink to k3s, already exists
[INFO]  Creating killall script /usr/local/bin/k3s-killall.sh
[INFO]  Creating uninstall script /usr/local/bin/k3s-uninstall.sh
[INFO]  env: Creating environment file /etc/systemd/system/k3s.service.env
[INFO]  systemd: Creating service file /etc/systemd/system/k3s.service
[INFO]  systemd: Enabling k3s unit
Created symlink /etc/systemd/system/multi-user.target.wants/k3s.service → /etc/systemd/system/k3s.service.
[INFO]  systemd: Starting k3s
Job for k3s.service failed because the control process exited with error code.
See "systemctl status k3s.service" and "journalctl -xe" for details.
```

**Expected behavior:**
k3s would become healthy and stable on version 1.19.1.

**Actual behavior:**
k3s crashing repeatedly.

**Additional context / logs:**

```
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: k3s.service: Found left-over process 2438 (containerd-shim) in control group while starting unit. Ignoring.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: This usually indicates unclean termination of a previous run, or service implementation deficiencies.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: k3s.service: Found left-over process 2447 (containerd-shim) in control group while starting unit. Ignoring.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: This usually indicates unclean termination of a previous run, or service implementation deficiencies.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: k3s.service: Found left-over process 2540 (containerd-shim) in control group while starting unit. Ignoring.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: This usually indicates unclean termination of a previous run, or service implementation deficiencies.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: k3s.service: Found left-over process 3458 (containerd-shim) in control group while starting unit. Ignoring.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: This usually indicates unclean termination of a previous run, or service implementation deficiencies.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: k3s.service: Found left-over process 3575 (containerd-shim) in control group while starting unit. Ignoring.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: This usually indicates unclean termination of a previous run, or service implementation deficiencies.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: k3s.service: Found left-over process 2488 (pause) in control group while starting unit. Ignoring.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: This usually indicates unclean termination of a previous run, or service implementation deficiencies.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: k3s.service: Found left-over process 2516 (pause) in control group while starting unit. Ignoring.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: This usually indicates unclean termination of a previous run, or service implementation deficiencies.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: k3s.service: Found left-over process 2631 (pause) in control group while starting unit. Ignoring.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: This usually indicates unclean termination of a previous run, or service implementation deficiencies.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: k3s.service: Found left-over process 2814 (coredns) in control group while starting unit. Ignoring.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: This usually indicates unclean termination of a previous run, or service implementation deficiencies.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: k3s.service: Found left-over process 2875 (metrics-server) in control group while starting unit. Ignoring.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: This usually indicates unclean termination of a previous run, or service implementation deficiencies.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: k3s.service: Found left-over process 3496 (pause) in control group while starting unit. Ignoring.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: This usually indicates unclean termination of a previous run, or service implementation deficiencies.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: k3s.service: Found left-over process 3602 (pause) in control group while starting unit. Ignoring.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: This usually indicates unclean termination of a previous run, or service implementation deficiencies.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: k3s.service: Found left-over process 3897 (entry) in control group while starting unit. Ignoring.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: This usually indicates unclean termination of a previous run, or service implementation deficiencies.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: k3s.service: Found left-over process 3917 (traefik) in control group while starting unit. Ignoring.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: This usually indicates unclean termination of a previous run, or service implementation deficiencies.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: k3s.service: Found left-over process 3988 (entry) in control group while starting unit. Ignoring.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: This usually indicates unclean termination of a previous run, or service implementation deficiencies.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: Starting Lightweight Kubernetes...
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: k3s.service: Found left-over process 2438 (containerd-shim) in control group while starting unit. Ignoring.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: This usually indicates unclean termination of a previous run, or service implementation deficiencies.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: k3s.service: Found left-over process 2447 (containerd-shim) in control group while starting unit. Ignoring.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: This usually indicates unclean termination of a previous run, or service implementation deficiencies.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: k3s.service: Found left-over process 2540 (containerd-shim) in control group while starting unit. Ignoring.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: This usually indicates unclean termination of a previous run, or service implementation deficiencies.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: k3s.service: Found left-over process 3458 (containerd-shim) in control group while starting unit. Ignoring.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: This usually indicates unclean termination of a previous run, or service implementation deficiencies.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: k3s.service: Found left-over process 3575 (containerd-shim) in control group while starting unit. Ignoring.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: This usually indicates unclean termination of a previous run, or service implementation deficiencies.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: k3s.service: Found left-over process 2488 (pause) in control group while starting unit. Ignoring.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: This usually indicates unclean termination of a previous run, or service implementation deficiencies.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: k3s.service: Found left-over process 2516 (pause) in control group while starting unit. Ignoring.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: This usually indicates unclean termination of a previous run, or service implementation deficiencies.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: k3s.service: Found left-over process 2631 (pause) in control group while starting unit. Ignoring.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: This usually indicates unclean termination of a previous run, or service implementation deficiencies.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: k3s.service: Found left-over process 2814 (coredns) in control group while starting unit. Ignoring.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: This usually indicates unclean termination of a previous run, or service implementation deficiencies.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: k3s.service: Found left-over process 2875 (metrics-server) in control group while starting unit. Ignoring.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: This usually indicates unclean termination of a previous run, or service implementation deficiencies.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: k3s.service: Found left-over process 3496 (pause) in control group while starting unit. Ignoring.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: This usually indicates unclean termination of a previous run, or service implementation deficiencies.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: k3s.service: Found left-over process 3602 (pause) in control group while starting unit. Ignoring.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: This usually indicates unclean termination of a previous run, or service implementation deficiencies.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: k3s.service: Found left-over process 3897 (entry) in control group while starting unit. Ignoring.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: This usually indicates unclean termination of a previous run, or service implementation deficiencies.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: k3s.service: Found left-over process 3917 (traefik) in control group while starting unit. Ignoring.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: This usually indicates unclean termination of a previous run, or service implementation deficiencies.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: k3s.service: Found left-over process 3988 (entry) in control group while starting unit. Ignoring.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: This usually indicates unclean termination of a previous run, or service implementation deficiencies.
Sep 22 17:35:21 ip-10-0-7-10 k3s[685]: time="2020-09-22T17:35:21Z" level=fatal msg="exec: \"k3s-server\": executable file not found in $PATH"
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: k3s.service: Main process exited, code=exited, status=1/FAILURE
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: k3s.service: Failed with result 'exit-code'.
Sep 22 17:35:21 ip-10-0-7-10 systemd[1]: Failed to start Lightweight Kubernetes.
```
Full log - [k3s.log.gz](https://github.com/rancher/k3s/files/5263154/k3s.log.gz)

**Comment by dnoland1** [comment]
Tried the same upgrade a second time on a new VM and it worked at expected, so must have hit an intermittent issue that is not consistently reproducible.

**Comment by brandond** [comment]
Were you by any chance low on disk space?

**Comment by dnoland1** [comment]
I think the chances of that are pretty low. I used a fresh Ubuntu VM with 10GB root disk and minimal install. Before k3s installation, there was almost 9GB free.
```
root@ip-10-0-7-10:~# df -h
Filesystem      Size  Used Avail Use% Mounted on
udev            1.9G     0  1.9G   0% /dev
tmpfs           389M  712K  389M   1% /run
/dev/nvme0n1p1  9.7G  805M  8.9G   9% /
tmpfs           1.9G     0  1.9G   0% /dev/shm
tmpfs           5.0M     0  5.0M   0% /run/lock
tmpfs           1.9G     0  1.9G   0% /sys/fs/cgroup
/dev/loop0       98M   98M     0 100% /snap/core/9993
/dev/loop1       29M   29M     0 100% /snap/amazon-ssm-agent/2012
tmpfs           389M     0  389M   0% /run/user/1000
```
I doubt running a vanilla k3s for 10-15 minutes and doing an upgrade will fill up that much space. But since I shutdown the environment, I guess I can't say with 100% certainty.

I ran through the steps in the description for a third time and did not hit the issue again. There was about 8GB free after the upgrade.

**Comment by dweomer** [comment]
@dnoland1 was there any process running in the background, i.e. `watch kubectl get pod`, which could have triggered the data-dir extraction while the k3s service was coming back up and hence likely also performing the extraction? I've noticed corruption resulting in segfaults and other odd behavior on systems with very slow disk i/o (most often for me, arm/arm64 on mini-sd) attributable, I think, to this bug: #1938

**Comment by dnoland1** [comment]
@dweomer That's an interesting question and theory. I think there's a good chance I was running `watch -n1 "kubectl get nodes; kubectl get pods -A"` in another terminal so I could monitor the upgrade process of the nodes and pods.

**Comment by erikwilson** [comment]
Hopefully this is fixed with https://github.com/rancher/k3s/pull/2349

**Comment by ShylajaDevadiga** [comment]
Tried to reproduced this issue upgrading from v1.18.3+k3s1 to v1.19.1+k3s1 while running watch -n1 "kubectl get nodes; kubectl get pods -A" from another terminal.
Upgrade was successful, Node was up and Pods in running running state. But during upgrade found fatal log messages
```
curl -sfL https://get.k3s.io | INSTALL_K3S_VERSION=v1.19.1+k3s1 sh -
[INFO]  Using v1.19.1+k3s1 as release
[INFO]  Downloading hash https://github.com/rancher/k3s/releases/download/v1.19.1+k3s1/sha256sum-amd64.txt
[INFO]  Downloading binary https://github.com/rancher/k3s/releases/download/v1.19.1+k3s1/k3s
[INFO]  Verifying binary download
[INFO]  Installing k3s to /usr/local/bin/k3s
[INFO]  Skipping /usr/local/bin/kubectl symlink to k3s, already exists
[INFO]  Skipping /usr/local/bin/crictl symlink to k3s, already exists
[INFO]  Skipping /usr/local/bin/ctr symlink to k3s, already exists
[INFO]  Creating killall script /usr/local/bin/k3s-killall.sh
[INFO]  Creating uninstall script /usr/local/bin/k3s-uninstall.sh
[INFO]  env: Creating environment file /etc/systemd/system/k3s.service.env
[INFO]  systemd: Creating service file /etc/systemd/system/k3s.service
[INFO]  systemd: Enabling k3s unit
Created symlink /etc/systemd/system/multi-user.target.wants/k3s.service → /etc/systemd/system/k3s.service.
[INFO]  systemd: Starting k3s
Job for k3s.service failed because the control process exited with error code.
See "systemctl status k3s.service" and "journalctl -xe" for details.
```
Logs:
```
Oct 14 23:11:43 ip-172-31-7-248 k3s[38942]: time="2020-10-14T23:11:43Z" level=fatal msg="extracting data: symlink coreutils /var/lib/rancher/k3s/data/1933e1efa2359d4f4d58bafb81c9367f94d978df6fc26e281b9badd400300673-tmp/bin/tty: file exists"
```

Validated fix using commit id 6b11d860374f938623479b0d06aeb3c1ac87ce86
Upgrade from v1.18.3+k3s1 to commit 6b11d860374f938623479b0d06aeb3c1ac87ce86 was successul. Above fatal logs not present.
```
kubectl get nodes
NAME              STATUS   ROLES    AGE   VERSION
ip-172-31-10-88   Ready    master   37s   v1.18.3+k3s1

curl -sfL https://get.k3s.io | INSTALL_K3S_COMMIT=6b11d860374f938623479b0d06aeb3c1ac87ce86 sh -
[INFO]  Using commit 6b11d860374f938623479b0d06aeb3c1ac87ce86 as release
[INFO]  Downloading hash https://storage.googleapis.com/k3s-ci-builds/k3s-6b11d860374f938623479b0d06aeb3c1ac87ce86.sha256sum
[INFO]  Downloading binary https://storage.googleapis.com/k3s-ci-builds/k3s-6b11d860374f938623479b0d06aeb3c1ac87ce86
[INFO]  Verifying binary download
[INFO]  Installing k3s to /usr/local/bin/k3s
[INFO]  Skipping /usr/local/bin/kubectl symlink to k3s, already exists
[INFO]  Skipping /usr/local/bin/crictl symlink to k3s, already exists
[INFO]  Skipping /usr/local/bin/ctr symlink to k3s, already exists
[INFO]  Creating killall script /usr/local/bin/k3s-killall.sh
[INFO]  Creating uninstall script /usr/local/bin/k3s-uninstall.sh
[INFO]  env: Creating environment file /etc/systemd/system/k3s.service.env
[INFO]  systemd: Creating service file /etc/systemd/system/k3s.service
[INFO]  systemd: Enabling k3s unit
Created symlink /etc/systemd/system/multi-user.target.wants/k3s.service → /etc/systemd/system/k3s.service.
[INFO]  systemd: Starting k3s
root@ip-172-31-10-88:~#

kubectl get nodes
NAME              STATUS   ROLES    AGE   VERSION
ip-172-31-10-88   Ready    master   65m   v1.19.2+k3s-6b11d860
```

**Comment by HidingCherry** [comment]
fyi - this happened to me when the /var/lib/rancher directory was on a separate device and mounted with "nodev,noexec".
